app/javascript/lib/validation/validate-file-content.js

```diff
@@ -276,99 +279,138 @@ function validateClusterCoordinates(headers) {
-  let issues = []
-  let table = [[]]
-  let delimiter = null
-
-  const { lines, mimeType } = await readLinesAndType(file)
-  const headerLines = lines.slice(0, 2)
-  issues = validateGzipEncoding({ fileName: file.name, lines, mimeType })
 
-  if (issues.length) {
-    return { table, issues, delimiter }
+/** reads the file and returns a fileInfo object along with an array of issues */
+async function parseFile(file, fileType) {
+  const fileInfo = {
+    fileSize: file.size,
+    fileName: file.name,
+    linesRead: 0,
+    numColumns: null,
+    fileMimeType: file.type,
+    fileType,
+    delimiter: null,
+    isGzipped: null
   }
+  const parseResult = { fileInfo, issues: [] }
+  try {
+    fileInfo.isGzipped = await validateGzipEncoding(file)
 
-  if (!file.name.endsWith('.gz')) {
-    if (['Cluster', 'Metadata'].includes(fileType)) {
-      if (lines.length < 2) {
-        return { table, delimiter, issues: [['error', 'cap:format:no-newlines', 'File does not contain newlines to separate rows']] }
-      }
-      // if there are no encoding issues, and this isn't a gzipped file, validate content
-      delimiter = sniffDelimiter(headerLines, mimeType)
-      table = []
-      for (let i = 0; i < lines.length; i++) {
-        table.push(lines[i].split(delimiter))
-      }
-      const headerTable = table.slice(0, 2)
-      issues = await validateCapFormat(headerTable, fileType)
-      issues = issues.concat(validateUniqueCellNamesWithinFile(table))
+    // if the file is compressed or we can't figure out the compression, don't try to parse further
+    if (fileInfo.isGzipped || !PARSEABLE_TYPES.includes(fileType)) {
+      return { fileInfo, issues: [] }
+    }
+    const parseFunctions = {
+      'Cluster': parseClusterFile,
+      'Metadata': parseMetadataFile
+    }
+    if (parseFunctions[fileType]) {
+      const chunker = new ChunkedLineReader(file)
+      const { issues, delimiter, numColumns } = await parseFunctions[fileType](chunker, fileInfo.fileMimeType)
+      fileInfo.linesRead = chunker.linesRead
+      fileInfo.delimiter = delimiter
+      fileInfo.numColumns = numColumns
+      parseResult.issues = issues
+    }
+  } catch (error) {
+    // get any unhandled or deliberate short-circuits
+    if (error instanceof ParseException) {
+      parseResult.issues.push(['error', error.key, error.message])
+    } else {
+      parseResult.issues.push(['error', 'parse:unhandled', error.message])
     }
   }
-  return { table, delimiter, issues }
+  return parseResult
 }
 
-/** Validate a local file, return list formatted of any detected errors */
+/** Validate a local file, return { errors, summary } object, where errors is an array of errors, and summary
+ * is a message like "Your file had 2 errors"
+ */
 export async function validateFileContent(file, fileType) {
-  const { table, delimiter, issues } = await parseFile(file, fileType)
+  const { fileInfo, issues } = await parseFile(file, fileType)
```

Comment:
Blocker: Towards performance tracking:

```suggestion
  const startTime = performance.now()
  const { fileInfo, issues } = await parseFile(file, fileType)
```